### DISCUSSION TITLE
From David Weaver: What is the proper (canonical) way to include the U extension, when listing the ISA of an implementation?

### DISCUSSION THREAD

**Discussion by rpsene** [riscv/riscv-isa-manual#987] [Q&A]
Section 27.11 of the RV Unpriv spec clearly specifies the canonical order of the ISA extensions.  It even includes the “S” and “H” extensions.  …However, oddly, it does not list the “U” extension (User mode).

What is the proper (canonical) way to include the U extension, when listing the ISA of an implementation?   It seems that the ISA string for an RV32GC implementation with U (RV32GCU?) should be different from one that does not include U (RV32GC).  

I couldn’t find that information in either the Unpriv or Priv specs, and don’t recall seeing it elsewhere. 
Any pointers where to find it would be much appreciated!

**Comment by rpsene** [comment]
Reply from @aswaterman:

The naming chapter in the unpriv spec doesn't actually define "S" and "H" extensions; rather, it defines S and H as prefixes for multi-letter extension names that pertain to privileged-architecture features.

I don't believe there is a canonical way to indicate user-mode support within an ISA string.  I've seen vendors indicate privileged-architecture support separate from ISA string, e.g., "an RV32GC implementation with M and U privilege modes".

The riscv-isa-manual github repo issue tracker is a good place for these questions, BTW.  It spreads the load of answering them (and might result in better answers than mine).

**Comment by Ganesh-403** [comment]
The short answer is: **there isn't a canonical way to encode support for U-mode in the ISA string.**

The ISA string is intended to describe the **instruction set extensions**, not the supported privilege modes. The ISA naming convention defines the ordering of standard instruction-set extensions (e.g., `IMAFDC`, `Z*`, `S*`, `H*`, etc.), but it does **not** define a single-letter `U` extension representing User mode. In fact, the `S` and `H` entries in the naming convention refer to the prefixes for *multi-letter privileged extensions* (e.g., `Ss*`, `Sh*`, `Sm*`), not standalone privilege modes. :contentReference[oaicite:0]{index=0}

This is also reflected in Linux's interpretation of ISA strings, which follows the specification's canonical ordering for instruction-set extensions but does not introduce a `U` extension in the ISA string. :contentReference[oaicite:1]{index=1}

So, for your example:

- **Not recommended:** `RV32GCU`
- **Recommended:** Keep the ISA string as `RV32GC` (or whatever extensions are implemented), and document the supported privilege modes separately.

For example:

> RV32GC implementation supporting Machine and User privilege modes.

or

> ISA: `RV32GC`  
> Privilege modes: M, U

This is consistent with Andrew Waterman's guidance in the linked reply: support for privilege modes is typically described separately from the ISA string, rather than by inventing a `U` suffix.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
